cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,253 +214,249 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
+ *
+ * Converts [first_bit_index, last_bit_index) to [first_word_index,
+ * last_word_index). The flag `end_of_segment` indicates whether the bit is at
+ * the end of a segment, in which case the word index should be incremented for
+ * bits at the start of a word.
+ */
+struct bit_to_word_index {
+  bit_to_word_index(bool end_of_segment) : end_of_segment(end_of_segment) {}
+  CUDA_DEVICE_CALLABLE size_type operator()(const size_type& bit_index) const
   {
-    auto bit_index = _d_bit_indices[i];
-    return word_index(bit_index) + ((_inclusive || intra_word_index(bit_index) == 0) ? 0 : 1);
+    return word_index(bit_index) + ((!end_of_segment || intra_word_index(bit_index) == 0) ? 0 : 1);
   }
+  bool const end_of_segment;
 };
 
 /**
  * @brief Functor that returns the number of set bits for a specified word
  * of a bitmask array.
  *
  */
-struct word_num_set_bits_functor {
-  word_num_set_bits_functor(bitmask_type const* bitmask_) : bitmask(bitmask_) {}
-  __device__ size_type operator()(size_type i) const
+struct count_set_bits_in_word {
+  count_set_bits_in_word(bitmask_type const* bitmask) : bitmask(bitmask) {}
+  CUDA_DEVICE_CALLABLE size_type operator()(size_type i) const
   {
     return static_cast<size_type>(__popc(bitmask[i]));
   }
   bitmask_type const* bitmask;
 };
 
-/**
- * @brief Given a bitmask, counts the number of set (1) bits in every range
- * `[indices_begin[2*i], indices_begin[(2*i)+1])` (where 0 <= i < std::distance(indices_begin,
- * indices_end) / 2).
- *
- * Returns an empty vector if `bitmask == nullptr`.
- *
- * @throws cudf::logic_error if `std::distance(indices_begin, indices_end) % 2 != 0`
- * @throws cudf::logic_error if `indices_begin[2*i] < 0 or indices_begin[2*i] >
- * indices_begin[(2*i)+1]`
- *
- * @param bitmask Bitmask residing in device memory whose bits will be counted
- * @param indices_begin An iterator representing the beginning of the range of indices specifying
- * ranges to count the number of set bits within
- * @param indices_end An iterator representing the end of the range of indices specifying ranges to
- * count the number of set bits within
- * @param streaam CUDA stream used for device memory operations and kernel launches
- *
- * @return A vector storing the number of non-zero bits in the specified ranges
- */
-template <typename IndexIterator>
-std::vector<size_type> segmented_count_set_bits(bitmask_type const* bitmask,
-                                                IndexIterator indices_begin,
-                                                IndexIterator indices_end,
-                                                rmm::cuda_stream_view stream)
+// Count set/unset bits in a segmented null mask, using offset iterators accessible by the device.
+template <typename OffsetIterator>
+rmm::device_uvector<size_type> segmented_count_bits_device(bitmask_type const* bitmask,
+                                                           size_type num_ranges,
```

Comment:
`num_ranges` effectively makes this an `*_n` algorithm. For consistency with most other iterator based algorithms in libcudf, consider taking a begin/end iterator for `first_bit_indices` and then just a begin iterator for `last_bit_indices`. `num_ranges` can then be computed from `distance(begin,end)`. 

This would also allow this function to be a unique overload of `segmented_count_bits` and no longer need the different name.

Reply:
Good suggestion. Done in 1f88efdb0c8a6406d510cc979aa85ec03226bc01.